File: RPi Experiments/RPi2-1/C/09_potFlowingLeds.c
#include <wiringPi.h>
#include <stdio.h>
#include<pcf8591.h>

#define PCF 120 //base pin for pcf8591

//turn on led_n
void led_on(int n)
{
	digitalWrite(n, LOW);
}

//turn off led_n
void led_off(int n)
{
	digitalWrite(n, HIGH);
}

int main(void)
{
    int value;
    int i;
    int pins[8] = {7, 0, 2, 3, 1, 4, 5, 6}; //wiringPi pin numbers for the LEDs


    if(wiringPiSetup() == -1)
    { 
		printf("setup wiringPi failed !");
		return 1; 
	}
      
    pcf8591Setup(PCF, 0x48); //setup pcf8591 on address 0x48

    for(i=0;i<8;i++) //iterate trough all pins
    {       
        pinMode(pins[i], OUTPUT); //initialize the LED pin as output
        led_off(pins[i]);
	}

    while(1)   
    {
        //iterate over the pins
        //turn the led on from lowest to the highest
        for(i=0;i<8;i++)
        {   
            value = analogRead(PCF + 0); //read the value from potentiometer
            analogWrite(PCF + 0, value);
			led_on(pins[i]);
			delay(value); //wait for ms
        }
        //fade from the highest to the lowest
        for(i=8;i>=0;i--)
        {  
            value = analogRead(PCF + 0); //read the value from potentiometer
            analogWrite(PCF + 0, value);
            led_off(pins[i]);
			delay(value); //wait for ms			
        }
        //turn the led on from highest to the lowest
        for(i=8;i>=0;i--)
        {  
            value = analogRead(PCF + 0); //read the value from potentiometer
            analogWrite(PCF + 0, value);
            led_on(pins[i]);
			delay(value); //wait for ms		
        }
        //fade from the lowest to the highest
        for(i=0;i<8;i++)
        {   
            value = analogRead(PCF + 0); //read the value from potentiometer
            analogWrite(PCF + 0, value);
			led_off(pins[i]);
			delay(value); //wait for ms
        }
	}

	return 0;
}

File: RPi Experiments/RPi2-1/Python/09_potFlowingLeds.py
import RPi.GPIO as GPIO
import time
import PCF8591 as ADC

pins = [7, 11, 13, 15, 12, 16, 18, 22]          # physical pin numbers for the LEDs
invertedPins = [22, 18, 16, 12, 15, 13, 11, 7]  # physical pin numbers for the LEDs in reverse order

def setup():
    GPIO.setwarnings(False)         # don't show warnings if GPIO is in use
    GPIO.setmode(GPIO.BOARD)        # use GPIO by physical location
    for pin in pins:                # iterate throug all pin numbers
        GPIO.setup(pin, GPIO.OUT)   # initialize the LED pin as output
        GPIO.output(pin, GPIO.HIGH) # turn off LED
    ADC.setup(0x48)

def loop():
    while True:
        for pin in pins:                # turn on LEDs from left to right
            value = ADC.read(0)         # read the potentiometer value
            ms = float(value) / 1000 
            ADC.write(ms)
            GPIO.output(pin, GPIO.LOW)  # turn on the LED	
            time.sleep(ms)              # wait for the number of seconds in ms
        for pin in invertedPins:        # turn off LEDs from right to left
            value = ADC.read(0)         # read the potentiometer value
            ms = float(value) / 1000 
            ADC.write(ms)
            GPIO.output(pin, GPIO.HIGH) # turn off the LED
            time.sleep(ms)              # wait for the number of seconds in ms
        for pin in invertedPins:        # turn on LEDs from right to left
            value = ADC.read(0)         # read the potentiometer value
            ms = float(value) / 1000
            ADC.write(ms)
            GPIO.output(pin, GPIO.LOW)  # turn on the LED
            time.sleep(ms)              # wait for the number of seconds in ms
        for pin in pins:                # turn off LEDs from left to right
            value = ADC.read(0)         # read the potentiometer value
            ms = float(value) / 1000
            ADC.write(ms)
            GPIO.output(pin, GPIO.HIGH) # turn off the LED
            time.sleep(ms)              # wait for the number of seconds in ms

def destroy():
	for pin in pins:
		GPIO.output(pin, GPIO.HIGH)    # turn off all LEDs
	GPIO.cleanup()                     # release resource

if __name__ == '__main__':     # program start from here
	setup()
	try:
		loop()
	except KeyboardInterrupt:  # when 'Ctrl+C' is pressed, the child program destroy() will be  executed.
		destroy()
